_docs/fact_type_fn_perf.md

```diff
@@ -0,0 +1,82 @@
+
+{% highlight clojure %}
+(defrecord IdHolder [id])
+(defquery even-ids []
+     [?even-ids <- (acc/all) :from IdHolder (even? id)])
+(defquery odd-ids []
+     [?odd-ids <- (acc/all) :from IdHolder (odd? id)])
+{% endhighlight %}
+
+Suppose that we insert (->IdHolder 1) into the session.  Clara will make the calls (even? 1) and (odd? 1) to determine which of these queries the IdHolder should match.  However, since a number cannot be both even and odd, it shouldn't be necessary to do this.  In most cases simply duplicating the number of calls won't be a major issue, but suppose that instead we had a thousand different rules that asserted that the ID was a specific value.  Something like
+
+{% highlight clojure %}
+(defrule rule1
+    [IdHolder (= id "A")])
+	=>
+	;; do something
+	)
+	
+(defrule rule2
+    [IdHolder (= id "AA")]
+	=>
+	;; do something
+	)
+	
+(defrule rule3
+    [IdHolder (= id "AAA")]
+	=>
+	;; do something
+	)
+.....
+(defrule rule1000
+     [IdHolder (clojure.string/join (repeat 1000 "A"))]
+	 =>
+	 ;; do something
+	 )
+{% endhighlight %}	 
+
+Where the elipses represent rules that are missing but follow the pattern laid out.  In this case, the id from each IdHolder would be extracted and compared for equality with 1000 diifferent strings, but since these conditions are mutually exclusive this shouldn't be necessary.  The reason for this comparison is that every fact is evaluated against every condition for which its fact type makes it eligible; Clara has no way of knowing that any given IdHolder can match at most one of these rules.  However, given a fact, Clara can determine which conditions its fact type makes it eligible for in constant time in addition to the time used by the :fact-type-fn and :ancestors-fn.  General discussion of this functionality can be found on [the fact type customization page](/docs/fact_type_customization).  Thus, in the following example, we will avoid the previous thousand equality checks per IdHolder fact.
+
+{% highlight clojure %}
+(defrule rule1
+      [[IdHolder "A"]]
+	  =>
+	  ;; do something
+	  )
+(defrule rule2
+      [[IdHolder "AA"]]
+	  =>
+	  ;; do something
+	  )
+(defrule rule3
+    [[IdHolder "AAA")]
+	=>
+	;; do something
+	)
+......
+(defrule rule1000
+     [[IdHolder (clojure.string/join (repeat 1000 "A")) ]]
+	 =>
+	 ;; do something
+	 )
+(mk-session :fact-type-fn (fn [fact]
+                                                 (cond 
+												        (instance? IdHolder fact) [IdHolder (:id fact)]
+														:else (type fact))))
+{% endhighlight %}	 
+
+Clara sessions can be roughly though of as containing a map from fact types to conditions that facts of that type are eligible to meet.  This custom :fact-type-fn function will determine the type of each IdHolder fact for rule processing services and then look that type up in this map, rather than essentially iterating over all of the map values to find which ones match.
```

Comment:
Ah, I see what you're saying now.  I think "condition" is probably better than "constraint" in this context because when the AlphaNode is activated with a fact that fact is then (possibly transitively) evaluated for all of the constraints in the condition.  There are, as you say, more precise ways we could parse this out but that seems like unnecessary complication of the subject of this page.  I'm a bit torn between saying "condition" and "rule" actually but at the moment I'm still leaning toward "condition" between the two.